test/e2e/util/kind/setup.go

```diff
@@ -0,0 +1,144 @@
+func (c *Cluster) Setup() {
+	var err error
+	c.tmpDir, err = ioutil.TempDir("", "kind-home")
+	gomega.Expect(err).To(gomega.BeNil())
+	fmt.Fprintf(ginkgo.GinkgoWriter, "creating Kind cluster named %q\n", c.Name)
+	c.run(exec.Command(*kindBinary, "create", "cluster", "--name", c.Name))
+	path := c.runWithOutput(exec.Command(*kindBinary, "get", "kubeconfig-path", "--name", c.Name))
+	c.kubepath = strings.TrimSpace(string(path))
+	fmt.Fprintf(ginkgo.GinkgoWriter, "kubeconfig path: %q. Can use the following to access the cluster:\n", c.kubepath)
+	fmt.Fprintf(ginkgo.GinkgoWriter, "export KUBECONFIG=%s\n", c.kubepath)
+}
+
+// Teardown attempts to delete the KIND cluster
+func (c *Cluster) Teardown() {
+	c.run(exec.Command(*kindBinary, "delete", "cluster", "--name", c.Name))
+	os.RemoveAll(c.tmpDir)
+}
+
+// LoadImage loads the specified image archive into the kind cluster
+func (c *Cluster) LoadImage(image string) {
+	fmt.Fprintf(
+		ginkgo.GinkgoWriter,
+		"loading image %q into Kind node\n",
+		image)
+	c.run(exec.Command(*kindBinary, "load", "docker-image", "--name", c.Name, image))
+}
+
+// ApplyYAML applies the provided manifest to the kind cluster
+func (c *Cluster) ApplyYAML(manifestPath string) {
+	c.run(exec.Command(
+		*kubectlBinary,
+		"create",
+		"--kubeconfig="+c.kubepath,
+		"-f", manifestPath,
+	))
+}
+
+// RestConfig returns a rest configuration pointed at the provisioned cluster
+func (c *Cluster) RestConfig() *restclient.Config {
+	cfg, err := clientcmd.BuildConfigFromFlags("", c.kubepath)
+	gomega.ExpectWithOffset(1, err).To(gomega.BeNil())
+	return cfg
+}
+
+// KubeClient returns a Kubernetes client pointing at the provisioned cluster
+func (c *Cluster) KubeClient() kubernetes.Interface {
+	cfg := c.RestConfig()
+	client, err := kubernetes.NewForConfig(cfg)
+	gomega.ExpectWithOffset(1, err).To(gomega.BeNil())
+	return client
+}
+
+func (c *Cluster) runWithOutput(cmd *exec.Cmd) []byte {
+	var stdout bytes.Buffer
+	cmd.Stdout = &stdout
+	c.run(cmd)
+	return stdout.Bytes()
+}
+
+func (c *Cluster) run(cmd *exec.Cmd) {
+	errPipe, err := cmd.StderrPipe()
+	gomega.ExpectWithOffset(1, err).To(gomega.BeNil())
+
+	cmd.Env = append(
+		cmd.Env,
+		// KIND positions the configuration file relative to HOME.
+		// To prevent clobbering an existing KIND installation, override this
+		// n.b. HOME isn't always set inside BAZEL
+		fmt.Sprintf("HOME=%s", c.tmpDir),
+		//needed for Docker. TODO(EKF) Should be properly hermetic
+		fmt.Sprintf("PATH=%s", os.Getenv("PATH")),
+	)
+
+	// Log output
+	go captureOutput(errPipe, "stderr")
+	if cmd.Stdout == nil {
+		outPipe, err := cmd.StdoutPipe()
+		gomega.ExpectWithOffset(1, err).To(gomega.BeNil())
+		go captureOutput(outPipe, "stdout")
+	}
+
+	gomega.ExpectWithOffset(1, cmd.Run()).To(gomega.BeNil())
+}
+
+func captureOutput(pipe io.ReadCloser, label string) {
+	buffer := &bytes.Buffer{}
+	defer pipe.Close()
+	for {
+		n, _ := buffer.ReadFrom(pipe)
```

Comment:
should this be reading until io.EOF or is n==0 the right thing?

Reply:
Oy, this rabbit hole led me down another, which is that we shouldn't be using exec.Cmd.Run with pipes, and instead should be using Start/Wait...  working on cleaning up.

Reply:
Updated this, but still need to test the changes.

Reply:
hold off on reviewing the changes, it won't work as is :joy:

Reply:
not a problem. i'm just starting to work on some of the trickier technical bits of the test proposal and this PR is a fairly crucial aspect :D